### PR TITLE
stm32f7: pwr: added basic support for pwr (VOS and overdrive)

### PR DESCRIPTION
**Pull request by matthewlai**
This hasn't been tested yet. I'll test it over the weekend.

### CODE REVIEW

**Comment by matthewlai** [comment]
This has been tested (chip boots up, and LED blinks at the right speed for 216 MHz).

File: include/libopencm3/stm32/f7/pwr.h [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+@author @htmlonly &copy; @endhtmlonly 2017 Matthew Lai <m@matthewlai.ca>
+
+@date 12 March 2017
+
+LGPL License Terms @ref lgpl_license
+ */
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2011 Stephen Caudle <scaudle@doceme.com>
```
Comment (karlp) [inline include/libopencm3/stm32/f7/pwr.h]: is there anything actually relevant here of their code remaining in any form?

File: include/libopencm3/stm32/f7/pwr.h [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+
+/* WUPF1: Wakeup Pin flag for PA0 */
+#define PWR_CSR2_WUPF1			(1 << 0)
+
+/* --- Function prototypes ------------------------------------------------- */
+
+enum pwr_vos_scale {
+	PWR_SCALE1, // <= 180MHz w/o overdrive, <= 216MHz w/ overdrive
+	PWR_SCALE2, // <= 168MHz w/o overdrive, <= 180MHz w/ overdrive
+	PWR_SCALE3, // <= 144MHz
```
Comment (karlp) [inline include/libopencm3/stm32/f7/pwr.h]: you can put inline /** comments */ here and doxygen will pick them as commentary for the enum values.

File: include/libopencm3/stm32/f7/pwr.h [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+};
+
+BEGIN_DECLS
+
+void pwr_set_vos_scale(enum pwr_vos_scale scale);
```
Comment (karlp) [inline include/libopencm3/stm32/f7/pwr.h]: Is this the same function name and enum name used in the other families?  (L0/L1/L4 to memory all have range scales)

File: include/libopencm3/stm32/f7/pwr.h (hunk 1 above)
Comment (matthewlai) [inline include/libopencm3/stm32/f7/pwr.h]: I don't think so. Removed.

File: include/libopencm3/stm32/f7/pwr.h (hunk 2 above)
Comment (matthewlai) [inline include/libopencm3/stm32/f7/pwr.h]: Done.

File: include/libopencm3/stm32/f7/pwr.h (hunk 3 above)
Comment (matthewlai) [inline include/libopencm3/stm32/f7/pwr.h]: Yeap! Copied from there.

**Comment by matthewlai** [comment]
I just noticed there's a problem with the rcc_clock_setup_hse_3v3() function. It works fine if I copy the function into my code, but not when it's compiled into the library (frequency is wrong).

I am still debugging, but looking at the disassembly, it looks like rcc_osc_on() is inlined when this is part of the library (since it happens to be in the same file). That may or may not be the reason. Maybe there's some kind of delay required somewhere.

**Comment by matthewlai** [comment]
Mystery solved. This PR is fine.

The problem is on F7 flash is mapped into two different regions - 0x08000000 like before (access through AXIM->AHB), and 0x00200000 (access through a dedicated ITCM bus, with ART cache). I was running through AXIM->AHB, without CPU caches enabled, which apparently made instruction alignment very important.

System clock rate was verified using a scope with MCO2 (clock-out from the chip).

I'll fix devices.data in another PR.

**Comment by karlp** [comment]
this landed in 17553da sorry it didn't get closed at the time

**Pull request closed without merging**